README.md
# toget

[![npm version](https://badge.fury.io/js/toget.svg)](https://badge.fury.io/js/toget) [![Build Status](https://travis-ci.org/protocoolmx/toget.svg?branch=master)](https://travis-ci.org/protocoolmx/toget)

Simple Node.js RESTful API builder for [request](https://github.com/request/request).

## Install

```
$ npm install toget --save
$ npm install request --save
```

**Note:** [request](https://github.com/request/request) 2+ required.

## Overview

**toget** makes available through its API the most used properties of [request](https://github.com/request/request).

```javascript
const toget = require('toget')(require('request'));

// All request made with `request` instance will have
// `http://domain.com` as base URL.
let request = toget('http://domain.com');

request('/user/:id', { id: 1 })
.then((response) => {
  console.log(response.body);
})
.catch((err) => {
  console.trace(err);
});
```

## Request API

`Request` is the class in charge to build and execute requests.

```javascript
const Request = require('toget/lib/Request');
```

#### constructor(base)

Creates `Request` instance.

+ `base` (String) - **protocol+auth+hostname+port** to be taken as base for request.

```javascript
let request = new Request('http://localhost:3000');

// Set request module for API to work.
request.upon(require('request'));
```

#### .get(path, params)

Set request `method`, `path` and optional `params`.

+ `path` (String) - The pathname property consists of the entire path section of the URL.
+ `params` (Object) - Possible object to replace values in path.

**Note:** The same for `.post()`, `.put()` and `.delete()` methods.

```javascript
request.get('/user/:id', { id: 1 })
```

#### .query(qs)

Used to specify query string with an Object.

+ `qs` (Object) - Object to be parsed into a URL query string.

```javascript
request.query({ sort: 'createdAt DESC' })
```

#### .headers(values)

Set HTTP request headers.

+ `values` (Object) - HTTP Headers as Object to be append in request.

```javascript
request.headers({ 'Accept-Encoding': 'gzip, deflate' })
```

#### .json()

Adds `Content-type: application/json` header and parses the response body as JSON.

```javascript
request.json()
```

#### .gzip()

Add an Accept-Encoding header to request compressed content encodings from the
server and decode supported content encodings in the response.

```javascript
request.gzip()
```

#### .jar(value)

Define your custom cookie jar.

```javascript
const url = 'http://localhost';
const req = new Request(url);
const j = req.request.jar();
j.setCookie(req.request.cookie('key1=value1'), url);

request.jar(j)
```

#### .timeout(value)

Integer containing the number of milliseconds to wait for a server to send response
before aborting the request.

```javascript
request.timeout(1500)
```

#### .body(data)

Entity body for POST and PUT requests.

+ `data` (Buffer|ReadStream|String|Object) - Entity body.

**Note:** If json is true, then body must be a JSON-serializable object.

```javascript
request.body({ firstName: 'Foo', lastName: 'Bar', age: 25 })
```

#### .encoding(value)

Set encoding code to the response body.

+ `value` (Object) - Represent the enconding code.

**Note:** If value is `undefined` means this is 'utf-8' by default. In the case for `null` is used as encoding code, 
the body is returned as a Buffer.

```javascript
request.encoding('utf-8')
```

#### .toPromise()

Make request and returns Promise instance to handle response and possible errors.

```javascript
request.toPromise()
.then((response) => {
  console.log(response.body);
})
.catch(console.error);
```

**Note:** `response` resolved in promise is an instance of `Response`.

#### .toOptions()

Returns request options Object ready to be used in request(options).

```javascript
request.get('/user').toOptions(); // { method: 'GET', url: 'http://localhost:3000/user' }
```

#### .exec(cb)

Shortcut for `request(req.get('/user').toOptions(), cb)` to use as a normal
request call. This method returns [request](https://github.com/request/request)
instance.

+ `cb` (Function) - Callback as specified in [request](https://github.com/request/request) docs; request(options, **callback**).

```javascript
request.get('/doodle.png').exec((err, response, body) => {
  if (err) {
    console.error(err);
    return;
  }

  // ...
})
.pipe(fs.createWriteStream('doodle.png'));
```

**Note:** The callback here is handled by request so this means `response` is
not the `response` you would expect returned by a promise but the
`response.rawResponse` instead.

## Response API

`Response` is the class in charge to handle response from request made.

```javascript
const Response = require('toget/lib/Response');
```

#### constructor(rawResponse)

Creates `Response` instance.

+ `rawResponse` (Object) - Response Object received in request callback.

```javascript
let request = new Response(rawResponse);
```

#### .body `Buffer|ReadStream|String`

Response body extracted from `rawResponse.body`.

```javascript
response.body
```

#### .headers `Object`

Response headers extracted from `rawResponse.headers`.

```javascript
response.headers
```

#### .statusCode `Number`

Response status code extracted from `rawResponse.statusCode`.

```javascript
response.statusCode
```

#### .url `String`

URL used in request.

```javascript
response.url
```

Method used in request (GET, POST, PUT, DELETE).

#### .method `String`

```javascript
response.method
```

#### .status `Object`

Response status code dictionary available as string-boolean as defined in [statusCodes.json](lib/statusCodes.json)

```javascript
response.status
```

#### .type `Number`

Response status range type code (1, 2, 3, 4, 5, etc).

```javascript
response.type
```

#### .info `Boolean`

Response status code range from 100 to 199.

```javascript
response.info
```

#### .ok `Boolean`

Response status code range from 200 to 299.

```javascript
response.ok
```

#### .clientError `Boolean`

Response status code range from 400 to 499.

```javascript
response.clientError
```

#### .serverError `Boolean`

Response status code range from 500 to 599.

```javascript
response.serverError
```

#### .error `Error|Object|Boolean`

Response error, false if no error or Error|Body instance otherwise.

```javascript
if (response.error) {
  throw response.error;
}
```

## Test

```
$ npm test
```

## License

MIT


CHANGELOG.md
# Change Log

### v0.5.0

 _2017-09-29_ [v0.4.0...v0.5.0](https://github.com/protocoolmx/toget/compare/v0.4.0...v0.5.0)

* [API BREAK CHANGE] `request` depedency is taken externally. [pull #15](https://github.com/protocoolmx/toget/pull/15).
* [CHANGE] eslint implemented [pull #13](https://github.com/protocoolmx/toget/pull/13).

### v0.4.0

 _2017-09-11_ [v0.3.0...v0.4.0](https://github.com/protocoolmx/toget/compare/v0.3.0...v0.4.0)

* [NEW OPTION] `jar` options implemented. [pull #11](https://github.com/protocoolmx/toget/pull/11).

### v0.3.0

 _2017-08-22_ [v0.2.0...v0.3.0](https://github.com/protocoolmx/toget/compare/v0.2.0...v0.3.0)

* [NEW OPTION] `timeout` option for requests added. [pull #8](https://github.com/protocoolmx/toget/pull/8).

### v0.2.0

 _2017-02-16_ [v0.1.2...v0.2.0](https://github.com/protocoolmx/toget/compare/v0.1.2...v0.2.0)

* [NEW OPTION] You can now set `encoding` option for body response. [767b9c7...beb674b](https://github.com/protocoolmx/toget/compare/767b9c7...beb674b).


test/setup.js
'use strict';

const jsonServer = require('json-server');
const fs = require('fs');

before('setup fake server', function (done) {
  // Always go back to original db.json
  fs.writeFileSync('test/fixtures/db.json', fs.readFileSync('test/fixtures/db.original.json'));

  const server = jsonServer.create();
  const router = jsonServer.router('test/fixtures/db.json');
  const middlewares = jsonServer.defaults();

  server.use(middlewares);

  // Add custom routes before JSON Server router
  server.get('/image', function (req, res) {
    res.set('content-type', 'image/png');
    fs.createReadStream('test/fixtures/image.png').pipe(res);
  });

  server.get('/timeout', function (req, res) {
    setTimeout(function () {
      res.statusCode = 200;
      res.setHeader('Content-Type', 'text/plain');
      res.end('NO_TIMEOUT');
    }, 2000);
  });

  server.get('/cookie', function (req, res) {
    res.statusCode = 200;
    res.setHeader('Content-Type', 'text/plain');
    res.cookie('key2', 'value2');
    res.end(req.headers.cookie);
  });

  server.use(router);
  server.listen(3000, () => done());
});


test/fixtures/tutils.js
'use strict';

const fs = require('fs');

module.exports = {
  db: () => {
    return new Promise((resolve, reject) => {
      fs.readFile('test/fixtures/db.json', 'utf-8', (err, data) => {
        if (err) {
          return reject(err);
        }

        return resolve(JSON.parse(data));
      });
    });
  }
};


test/eslint/eslint.test.js
'use strict';

const lint = require('mocha-eslint');

// Array of paths to lint
// Note: a seperate Mocha test will be run for each path and each file which
// matches a glob pattern
const paths = [
  'lib',
  'index.js',
];

const options = {
  // Specify style of output
  formatter: 'compact',  // Defaults to `stylish`

  // Only display warnings if a test is failing
  alwaysWarn: false,  // Defaults to `true`, always show warnings

  // Increase the timeout of the test if linting takes to long
  timeout: 5000,  // Defaults to the global mocha `timeout` option

  // Increase the time until a test is marked as slow
  slow: 1000,  // Defaults to the global mocha `slow` option

  // Consider linting warnings as errors and return failure
  strict: true,  // Defaults to `false`, only notify the warnings
};

// Run the tests
lint(paths, options);

lib/Request.js
'use strict';

const url = require('url');
const pathToRegexp = require('path-to-regexp');
const Response = require('./Response');

class Request {

  constructor(base) {
    if (!base) {
      throw new Error('"base" must be defined!');
    }

    // Create empty options object to be built for request.
    this._options = {};

    // Init urlObject from given base.
    this.urlObject = url.parse(base);

    // Override default values for urlObject.
    this.urlObject.hash = null;
    this.urlObject.search = null;
    this.urlObject.query = null;
    this.urlObject.pathname = '/';
    this.urlObject.path = '/';
    this.urlObject.href = null;
  }

  /**
   * @param {String} method - request http method.
   * @param {String} path - urlObject.path to be passed to url.format(urlObject).
   * @param {Object} obj - Path Object to parse with pathToRegexp.
   */
  _methodPath(method, path, obj) {
    this._options.method = method;
    this.urlObject.pathname = path;

    if (obj && typeof obj === 'object') {
      this.pathObject = obj;
    }
  }

  _buildOptions() {
    // Make sure to not overbuilt options Object.
    if (this._optionsBuilt) {
      return;
    }

    // If `this.pathObject` is defined then parse it against this.urlObject.pathname.
    if (this.pathObject) {
      const toPath = pathToRegexp.compile(this.urlObject.pathname);
      this.urlObject.pathname = toPath(this.pathObject);
    }

    // If method is not of POST or PUT type then remove body from options.
    if (!['POST', 'PUT'].some(method => method === this._options.method)) {
      delete this._options.body;
    }

    // Set request URL.
    this._options.url = url.format(this.urlObject);

    this._optionsBuilt = true;
  }

  _checkRequestModule() {
    if (!this.request) {
      throw Error('Request#upon() must be called first with request module');
    }
  }

  /**
   * Appen `request` module.
   *
   * @param {Object} request
   * @memberof Request
   */
  upon(request) {
    this.request = request;

    return this;
  }

  /**
   * Build request with options Object and returns Promise.
   *
   * @return {Promise}
   */
  toPromise() {
    this._checkRequestModule();

    // Make sure to not overdefined promise.
    if (!this._promise) {
      // Build request promise.
      this._promise = new Promise((resolve, reject) => {
        this.request(this.toOptions(), (err, response) => {
          if (err) {
            return reject(err);
          }

          // If url and method are not set on response then set them.
          response.url = response.url || this._options.url;
          response.method = response.method || this._options.method;

          return resolve(new Response(response));
        });
      });
    }

    return this._promise;
  }

  /**
   * @return {Object} Valid options Object ready to be passed in request.
   */
  toOptions() {
    if (!this._optionsBuilt) {
      this._buildOptions();
    }

    return this._options;
  }

  /**
   * Make a normal request with `this._options` and return request instance.
   *
   * @param {Function} [cb] - Callback for request.
   * @return {Object} request instance.
   */
  exec(cb) {
    this._checkRequestModule();

    return this.request(this.toOptions(), cb);
  }

  /**
   * Execute request, return promise and gets resolved with response.
   */
  then(cb) {
    return this.toPromise().then(cb);
  }

  /**
   * Execute request, returns a promise and gets rejected with error.
   */
  catch(cb) {
    return this.toPromise().catch(cb);
  }

  /**
   * Set request as GET and passes path.
   */
  get(path, obj) {
    this._methodPath('GET', path, obj);

    return this;
  }

  /**
   * Set request as POST and passes path.
   */
  post(path, obj) {
    this._methodPath('POST', path, obj);

    return this;
  }

  /**
   * Set request as PUT and passes path.
   */
  put(path, obj) {
    this._methodPath('PUT', path, obj);

    return this;
  }

  /**
   * Set request as DELETE and passes path.
   */
  delete(path, obj) {
    this._methodPath('DELETE', path, obj);

    return this;
  }

  /**
   * Query string to be build using: url.format(urlObject).
   *
   * @see https://nodejs.org/dist/latest-v4.x/docs/api/url.html
   */
  query(qs) {
    this.urlObject.query = qs;

    return this;
  }

  /**
   * Appends collection of headers to request options Object.
   *
   * @desc HTTP headers.
   *
   * @param {Object} values - key-value Object of headers.
   */
  headers(values) {
    this._options.headers = values || {};

    return this;
  }

  /**
   * Appends json as `true` to request options Object.
   *
   * @desc Sets body to JSON representation of value and adds
   * Content-type: application/json header. Additionally, parses the response
   * body as JSON.
   */
  json() {
    this._options.json = true;

    return this;
  }

  /**
   * Appends gzip as `true` to request options Object.
   *
   * @desc Add an Accept-Encoding header to request compressed content
   * encodings from the server and decode supported content encodings in the
   * response.
   *
   * Note: Automatic decoding of the response content is performed on the body
   * data returned through request (both through the request stream and passed
   * to the callback function) but is not performed on the response stream
   * (available from the response event) which is the unmodified
   * http.IncomingMessage object which may contain compressed data.
   *
   * @return {Request}
   */
  gzip() {
    this._options.gzip = true;

    return this;
  }

  /**
   * Integer containing the number of milliseconds to wait for a server to send
   * response before aborting the request.
   *
   * @param {number} value - Milliseconds before timing out request.
   * @return {Request}
   */
  timeout(value) {
    this._options.timeout = value;

    return this;
  }

  /**
   * Define your custom cookie jar.
   *
   * @param {object} value - Cookie jar returned by `request.jar()`.
   * @return {Request}
   */
  jar(value) {
    this._options.jar = value;

    return this;
  }

  /**
   * Appends body to request options Object.
   *
   * @param {Bufer|String|ReadStream} data - Entity body for POST and PUT
   * requests (If json is true, then body must be a JSON-serializable object).
   */
  body(data) {
    this._options.body = data;

    return this;
  }

  /**
   * Set encoding to response body.
   *
   * @param {Object} value - Code for encoding response body. If value is null,
   * the body is returned as a Buffer otherwise the response body is encoding
   * with the code from value.
   */
  encoding(value) {
    this._options.encoding = value;

    return this;
  }
}

module.exports = Request;


lib/request.js
'use strict';

const url = require('url');
const pathToRegexp = require('path-to-regexp');
const Response = require('./response');

class Request {

  constructor(base) {
    if (!base) {
      throw new Error('"base" must be defined!');
    }

    // Make 'request' available in Request class.
    this.request = require('request'); // eslint-disable-line

    // Create empty options object to be built for request.
    this._options = {};

    // Init urlObject from given base.
    this.urlObject = url.parse(base);

    // Override default values for urlObject.
    this.urlObject.hash = null;
    this.urlObject.search = null;
    this.urlObject.query = null;
    this.urlObject.pathname = '/';
    this.urlObject.path = '/';
    this.urlObject.href = null;
  }

  /**
   * @param {String} method - request http method.
   * @param {String} path - urlObject.path to be passed to url.format(urlObject).
   * @param {Object} obj - Path Object to parse with pathToRegexp.
   */
  _methodPath(method, path, obj) {
    this._options.method = method;
    this.urlObject.pathname = path;

    if (obj && typeof obj === 'object') {
      this.pathObject = obj;
    }
  }

  _buildOptions() {
    // Make sure to not overbuilt options Object.
    if (this._optionsBuilt) {
      return;
    }

    // If `this.pathObject` is defined then parse it against this.urlObject.pathname.
    if (this.pathObject) {
      const toPath = pathToRegexp.compile(this.urlObject.pathname);
      this.urlObject.pathname = toPath(this.pathObject);
    }

    // If method is not of POST or PUT type then remove body from options.
    if (!['POST', 'PUT'].some(method => method === this._options.method)) {
      delete this._options.body;
    }

    // Set request URL.
    this._options.url = url.format(this.urlObject);

    this._optionsBuilt = true;
  }

  /**
   * Build request with options Object and returns Promise.
   *
   * @return {Promise}
   */
  toPromise() {
    // Make sure to not overdefined promise.
    if (!this._promise) {
      // Build request promise.
      this._promise = new Promise((resolve, reject) => {
        this.request(this.toOptions(), (err, response) => {
          if (err) {
            return reject(err);
          }

          // If url and method are not set on response then set them.
          response.url = response.url || this._options.url;
          response.method = response.method || this._options.method;

          return resolve(new Response(response));
        });
      });
    }

    return this._promise;
  }

  /**
   * @return {Object} Valid options Object ready to be passed in request.
   */
  toOptions() {
    if (!this._optionsBuilt) {
      this._buildOptions();
    }

    return this._options;
  }

  /**
   * Make a normal request with `this._options` and return request instance.
   *
   * @param {Function} [cb] - Callback for request.
   * @return {Object} request instance.
   */
  exec(cb) {
    return this.request(this.toOptions(), cb);
  }

  /**
   * Execute request, return promise and gets resolved with response.
   */
  then(cb) {
    return this.toPromise().then(cb);
  }

  /**
   * Execute request, returns a promise and gets rejected with error.
   */
  catch(cb) {
    return this.toPromise().catch(cb);
  }

  /**
   * Set request as GET and passes path.
   */
  get(path, obj) {
    this._methodPath('GET', path, obj);

    return this;
  }

  /**
   * Set request as POST and passes path.
   */
  post(path, obj) {
    this._methodPath('POST', path, obj);

    return this;
  }

  /**
   * Set request as PUT and passes path.
   */
  put(path, obj) {
    this._methodPath('PUT', path, obj);

    return this;
  }

  /**
   * Set request as DELETE and passes path.
   */
  delete(path, obj) {
    this._methodPath('DELETE', path, obj);

    return this;
  }

  /**
   * Query string to be build using: url.format(urlObject).
   *
   * @see https://nodejs.org/dist/latest-v4.x/docs/api/url.html
   */
  query(qs) {
    this.urlObject.query = qs;

    return this;
  }

  /**
   * Appends collection of headers to request options Object.
   *
   * @desc HTTP headers.
   *
   * @param {Object} values - key-value Object of headers.
   */
  headers(values) {
    this._options.headers = values || {};

    return this;
  }

  /**
   * Appends json as `true` to request options Object.
   *
   * @desc Sets body to JSON representation of value and adds
   * Content-type: application/json header. Additionally, parses the response
   * body as JSON.
   */
  json() {
    this._options.json = true;

    return this;
  }

  /**
   * Appends gzip as `true` to request options Object.
   *
   * @desc Add an Accept-Encoding header to request compressed content
   * encodings from the server and decode supported content encodings in the
   * response.
   *
   * Note: Automatic decoding of the response content is performed on the body
   * data returned through request (both through the request stream and passed
   * to the callback function) but is not performed on the response stream
   * (available from the response event) which is the unmodified
   * http.IncomingMessage object which may contain compressed data.
   *
   * @return {Request}
   */
  gzip() {
    this._options.gzip = true;

    return this;
  }

  /**
   * Integer containing the number of milliseconds to wait for a server to send
   * response before aborting the request.
   *
   * @param {number} value - Milliseconds before timing out request.
   * @return {Request}
   */
  timeout(value) {
    this._options.timeout = value;

    return this;
  }

  /**
   * Define your custom cookie jar.
   *
   * @param {object} value - Cookie jar returned by `request.jar()`.
   * @return {Request}
   */
  jar(value) {
    this._options.jar = value;

    return this;
  }

  /**
   * Appends body to request options Object.
   *
   * @param {Bufer|String|ReadStream} data - Entity body for POST and PUT
   * requests (If json is true, then body must be a JSON-serializable object).
   */
  body(data) {
    this._options.body = data;

    return this;
  }

  /**
   * Set encoding to response body.
   *
   * @param {Object} value - Code for encoding response body. If value is null,
   * the body is returned as a Buffer otherwise the response body is encoding
   * with the code from value.
   */
  encoding(value) {
    this._options.encoding = value;

    return this;
  }
}

module.exports = Request;


lib/Response.js
'use strict';

const statusCodes = JSON.parse(require('fs').readFileSync(`${__dirname}/statusCodes.json`));

class Response {

  constructor(rawResponse) {
    // Make raw response available in response.
    this.rawResponse = rawResponse;

    // Get body, headers and statusCode exactly like it is in response.
    this.body = rawResponse.body;
    this.headers = rawResponse.headers;
    this.statusCode = rawResponse.statusCode;

    // Read dictionary of status codes from lib/statusCodes.json
    // and set false|true values depending on statusCode from response.
    this.status = {};
    Object.keys(statusCodes).forEach((code) => {
      this.status[statusCodes[code]] = this.statusCode === parseInt(code, 10);
    });

    // Make url and method used in request available in response.
    this.url = rawResponse.url;
    this.method = rawResponse.method;
  }

  /**
   * @return {Number} Status range type code (1, 2, 3, 4, 5, etc).
   */
  get type() {
    return Math.floor(this.statusCode / 100);
  }

  /**
   * Status code range from 100 to 199.
   *
   * @return {Boolean}
   */
  get info() {
    return this.type === 1;
  }

  /**
   * Status code range from 200 to 299.
   *
   * @return {Boolean}
   */
  get ok() {
    return this.type === 2;
  }

  /**
   * Status code range from 400 to 499.
   *
   * @return {Boolean}
   */
  get clientError() {
    return this.type === 4;
  }

  /**
   * Status code range from 500 to 599.
   *
   * @return {Boolean}
   */
  get serverError() {
    return this.type === 5;
  }

  /**
   * Response error, false if no error or Error|Object instance otherwise.
   *
   * @return {Error|Object|Boolean} Possible error data.
   */
  get error() {
    if (this.clientError || this.serverError) {
      if (this.body) {
        return this.body;
      }

      return new Error(`Response got ${statusCodes[this.statusCode] || this.type}`);
    }

    return false;
  }
}

module.exports = Response;


lib/response.js
'use strict';

const fs = require('fs');
const statusCodes = JSON.parse(require('fs').readFileSync(__dirname + '/statusCodes.json'));

class Response {

  constructor(rawResponse) {
    // Make raw response available in response.
    this.rawResponse = rawResponse;

    // Get body, headers and statusCode exactly like it is in response.
    this.body = rawResponse.body;
    this.headers = rawResponse.headers;
    this.statusCode = rawResponse.statusCode;

    // Read dictionary of status codes from lib/statusCodes.json
    // and set false|true values depending on statusCode from response.
    this.status = {};
    Object.keys(statusCodes).forEach((code) => {
      this.status[statusCodes[code]] = this.statusCode === parseInt(code);
    });

    // Make url and method used in request available in response.
    this.url = rawResponse.url;
    this.method = rawResponse.method;
  }

  /**
   * @return {Number} Status range type code (1, 2, 3, 4, 5, etc).
   */
  get type() {
    return this.statusCode / 100 | 0;
  }

  /**
   * Status code range from 100 to 199.
   *
   * @return {Boolean}
   */
  get info() {
    return this.type === 1;
  }

  /**
   * Status code range from 200 to 299.
   *
   * @return {Boolean}
   */
  get ok() {
    return this.type === 2;
  }

  /**
   * Status code range from 400 to 499.
   *
   * @return {Boolean}
   */
  get clientError() {
    return this.type === 4;
  }

  /**
   * Status code range from 500 to 599.
   *
   * @return {Boolean}
   */
  get serverError() {
    return this.type === 5;
  }

  /**
   * Response error, false if no error or Error|Object instance otherwise.
   *
   * @return {Error|Object|Boolean} Possible error data.
   */
  get error() {
    if (this.clientError || this.serverError) {

      if (this.body) {
        return this.body;
      }

      return new Error(`Response got ${statusCodes[this.statusCode] || this.type}`);
    }

    return false;
  }
}

module.exports = Response;
